### DISCUSSION TITLE
with_loader_criteria is not applied during join

### DISCUSSION THREAD

**Discussion by simon-sk** [sqlalchemy/sqlalchemy#10364] [Usage Questions]
We are seeing an unexpected behaviour with the with_loader_criteria where our filter set by the `do_orm_execute` event hook is not applied if we join multiple models. This is our simplified setup:

```python
from sqlalchemy import String, create_engine, event, orm, select
from sqlalchemy.orm import Mapped, mapped_column, Session, declarative_base, ORMExecuteState

engine = create_engine("postgresql://postgres:mysecretpassword@localhost:5432", echo=True)
session = Session(engine)
Base = declarative_base()


class Tenant:
    tenant_id: Mapped[str] = mapped_column("tenant_id", String(), primary_key=True)
    

class A(Base):
    __tablename__ = "a"

    id: Mapped[str] = mapped_column(primary_key=True)


class B(Tenant, Base):
    __tablename__ = "b"

    a_id: Mapped[str] = mapped_column()


@event.listens_for(Session, "do_orm_execute")
def _add_tenant_id_filtering_criteria_to_queries(execute_state: ORMExecuteState) -> None:
    """simplified filter to always filter for tenant '1'"""
    execute_state.statement = execute_state.statement.options(
        orm.with_loader_criteria(
            Tenant,
            lambda cls: cls.tenant_id == "1",
            propagate_to_loaders=True,
            track_closure_variables=True,
            include_aliases=True,
        )
    )

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

session.add(A(id="ABC"))
session.add(B(a_id="ABC", tenant_id="1"))
session.add(B(a_id="ABC", tenant_id="2"))
session.commit()
```

Querying `B` works as expected (the filter is added to the query and only one row is returned with tenant_id="1"):
```python
    result = session.execute(select(B))
    # Executed Statement:
    #  SELECT b.a_id, b.tenant_id
    #  FROM b
    #  WHERE b.tenant_id = '1'
    assert len(result.fetchall()) == 1  # Works, only one entity of class B returned with tenant_id="1"
```

However, when joining `A` with `B`, the loader criteria is not applied (the filter is not added to the query and both B entities are returned):
```python
    result = session.execute(select(A, B).join(B, (A.id == B.a_id)))
    # Executed Statement:
    #  SELECT a.id, b.a_id, b.tenant_id
    #  FROM a JOIN b ON a.id = b.a_id
    assert len(result.fetchall()) == 1  # Fails since both B(tenant_id="1") and B(tenant_id="2") are returned
```

Is it possible to also add the filter criteria to joins? From the [description](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#sqlalchemy.orm.with_loader_criteria) of `with_loader_criteria` I would expect that this condition is added as well:
```SQL
SELECT a.id, b.a_id, b.tenant_id
FROM a JOIN b ON a.id = b.a_id
WHERE b.tenant_id = '1'
```

A similar problem is also described here: https://stackoverflow.com/questions/74197458/how-can-i-get-with-loader-criteria-to-apply-to-multiple-models-in-a-query

**Comment by zzzeek** [comment]
pretty strange, this seems to be an intentional decision to not add extra criteria if the ON clause is spelled out as an expression already.  This turns off that decision:

```diff
diff --git a/lib/sqlalchemy/orm/util.py b/lib/sqlalchemy/orm/util.py
index f22a798803..f142f3b326 100644
--- a/lib/sqlalchemy/orm/util.py
+++ b/lib/sqlalchemy/orm/util.py
@@ -1884,7 +1884,8 @@ class _ORMJoin(expression.Join):
                     {"parententity": parententity}
                 )
 
-        augment_onclause = onclause is None and _extra_criteria
+        #augment_onclause = onclause is None and _extra_criteria
+        augment_onclause = bool(_extra_criteria)
         expression.Join.__init__(self, left, right, onclause, isouter, full)
 
         assert self.onclause is not None

```

So!  We make a bug and run through the test suite to see what this conditional had in mind

**Reply by simon-sk** [reply]
@zzzeek Thanks for the quick reply! Let me know if there is anything else you need from my side 👍

**Reply by zzzeek** [reply]
nope....it's #10365 and it is ..you know a rabbit hole.....different test cases showing different things....it'll work out

**Reply by jean-sh** [reply]
My team is having the same issue, we are on 1.4 and I see this has been fixed in 2.0 but not 1.4, any reason for that? Could this be backported to 1.4? I can open an issue and a PR with the same patch if necessary

**Reply by zzzeek** [reply]
we aren't generally doing releases of 1.4 and I would note nothing has been committed to 1.4 in awhile.  the issue here is not a regression and only shows up if someone writes new code, which is all the usual criteria we have for "not a backport"; we want people upgrading to the latest series if they are doing new development.

that said, this fix is very trivial so I put up a review for it at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5170 just to test, but we dont have much momentum for there to be a new 1.4 release.  1.4 still has python 2 code in it and it already is falling behind more recent flake8 rules that I had to fix there.

**Reply by jean-sh** [reply]
Thanks for your quick answer!
Migrating for us is definitely not trivial –we use old-style queries, i.e. `session.query(MyModel)...` everywhere in our code base– so we're stuck on 1.4 for the moment unfortunately.

If the fix can at least be committed, we could target the HEAD of 1.4 I guess 🤔

**Reply by jean-sh** [reply]
I'd like to add that this is a security vulnerability for us (in our application), as we use `with_loader_criteria` as a way to apply read permissions to our database objects; which is a use of the function explicitly stated in the documentation.

**Reply by zzzeek** [reply]
let's not go there (claiming SQLAlchemy has a CVE).    a CVE is when a program that runs completely correctly has an unseen window for an exploit for which the program has no reasonable ability to guard against.   in this case, the wrong SQL is emitted in your program and the query is wrong from the start; basic unit tests would reveal that the query doesnt work and a query that is filtering on roles is not functioning.

that is, this is a broken lock for you, it's not a lock that seems to work but is vulnerable.

**Reply by jean-sh** [reply]
I definitely wasn't trying to imply anything about a CVE, sorry if that's how that came across :)  I just wanted to emphasize that this was not a minor bug in our application and that having the fix in 1.4 would be very helpful

**Reply by CaselIT** [reply]
> Migrating for us is definitely not trivial –we use old-style queries, i.e. session.query(MyModel)... everywhere in our code base– so we're stuck on 1.4 for the moment unfortunately.

note that all this still works perfectly fine in 2.0 and even has typing support. I'm not sure where you get the impression that that would not work.

**Reply by zzzeek** [reply]
I've backported and released SQLAlchemy 1.4.52 with this fix alone: https://pypi.org/project/SQLAlchemy/1.4.52/

**Reply by jean-sh** [reply]
Thanks a lot! You've avoided us a very ugly monkeypatch (and we've put migrating to 2.0 on our radar 😉 )
